File: code/design-patterns/behavioral/memento/go/memento_usage.go
package main

import "fmt"

func main() {
	document := &TextDocument{content: ""}
	editor := NewTextEditor(document)

	editor.Write("Hello, ")
	editor.Save() // Save state to memento
	editor.Write("world!")
	fmt.Println("Current State:", editor.GetContent()) // Outputs: Hello, world!

	editor.Undo()                                          // Restore previous state from memento
	fmt.Println("Restored to State:", editor.GetContent()) // Outputs: Hello,

	editor.Undo()                                          // Attempt to restore previous state, but history is empty
	fmt.Println("Restored to State:", editor.GetContent()) // Outputs: Hello, (no change)
}

File: code/design-patterns/behavioral/memento/python/memento_usage.py
from memento import TextDocument, TextEditor


document = TextDocument("")
editor = TextEditor(document)

editor.write("Hello, ")
editor.save()  # Save state to memento
editor.write("world!")
print("Current State:", editor.get_content())  # Outputs: Hello, world!

editor.undo()  # Restore previous state from memento
print("Restored to State:", editor.get_content())  # Outputs: Hello,

editor.undo()  # Attempt to restore previous state, but history is empty
print("Restored to State:", editor.get_content())  # Outputs: Hello, (no change)
